liquidhaskell-boot/src-ghc/Liquid/GHC/API/Extra.hs

```diff
@@ -170,46 +170,43 @@ apiCommentsParsedSource ps =
     spanToLineColumn =
       fmap (\s -> (srcSpanStartLine s, srcSpanStartCol s)) . srcSpanToRealSrcSpan
 
--- | Adds NOINLINE pragmas to all local bindings in the module.
+-- | Adds NOINLINE pragmas to all bindings in the module.
 --
 -- This prevents the simple optimizer from inlining such bindings which might
 -- have specs that would otherwise be left dangling.
 --
 -- https://gitlab.haskell.org/ghc/ghc/-/issues/24386
 --
-addNoInlinePragmasToLocalBinds :: ParsedModule -> ParsedModule
-addNoInlinePragmasToLocalBinds ps =
-    ps { pm_parsed_source = go (pm_parsed_source ps) }
+addNoInlinePragmasToBinds :: TcGblEnv -> TcGblEnv
+addNoInlinePragmasToBinds tcg = tcg{ tcg_binds = go False (tcg_binds tcg) }
   where
-    go :: forall a. Data a => a -> a
-    go = gmapT ((id `extT` addNoInlinePragmas) . go)
-
-    addNoInlinePragmas :: HsValBinds GhcPs -> HsValBinds GhcPs
-    addNoInlinePragmas = \case
-      ValBinds x binds sigs ->
-          ValBinds x binds (newSigs ++ dropInlinePragmas sigs)
-        where
-          dropInlinePragmas = filter (not . isInlineSig . unLoc)
-
-          isInlineSig InlineSig{} = True
-          isInlineSig _ = False
-
-          newSigs = concatMap (traverse noInlinePragmaForBind) $ bagToList binds
-
-          noInlinePragmaForBind :: HsBindLR GhcPs GhcPs -> [Sig GhcPs]
-          noInlinePragmaForBind bndr =
-              [ InlineSig
-                  []
-                  (noLocA b)
-                  defaultInlinePragma
-                    { inl_inline = NoInline $ SourceText $ occNameFS $ rdrNameOcc b
-                    , inl_act = NeverActive
-                    }
-              | b <- collectHsBindBinders CollNoDictBinders bndr
-              ]
-
-      bnds -> bnds
-
+    go :: forall a. Data a => Bool -> a -> a
+    go isAbs = gmapT $ go isAbs `extT` markHsBind isAbs
+
+    markHsBind :: Bool -> HsBind GhcTc -> HsBind GhcTc
+    markHsBind isAbs = \case
+        VarBind ext var rhs -> VarBind ext (markId isAbs var) (go False rhs)
+        FunBind ext var matches -> FunBind ext (markId isAbs <$> var) (go False matches)
+        PatBind ext lhs mult rhs -> PatBind ext lhs mult (go False rhs)
+        PatSynBind{} -> error "markNoInline: PatSynBind"
```

Comment:
How is this error case avoided when the module has pattern synonyms?

And why can't patterns synonyms be ignored?

Reply:
The GHC desugarer already errors out on `PatSyBind`s -- they're not supposed to be here, since the typechecker replaces them with the binders for the matcher and the builder.

We could just ignore them, but that wouldn't be a "motivated ignoring". 

Best would be to change GHC so that `XPatSynBind GhcTc` is isomorphic to `Void`. I can prepare a patch tomorrow, but knowing GHC, it could take a long time until it lands in `master`.

Reply:
```suggestion
        PatSynBind{} -> error "markNoInline: unexpected PatSynBind, should have been eliminated by the typechecker"
```

Reply:
https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13441 FYI